source/server/overload_manager_impl.cc

```diff
@@ -0,0 +1,188 @@
+#include "server/overload_manager_impl.h"
+
+#include "common/common/fmt.h"
+#include "common/config/utility.h"
+#include "common/protobuf/utility.h"
+
+#include "server/resource_monitor_config_impl.h"
+
+namespace Envoy {
+namespace Server {
+
+namespace {
+
+class ThresholdTriggerImpl : public OverloadActionImpl::Trigger {
+public:
+  ThresholdTriggerImpl(const envoy::config::overload::v2alpha::ThresholdTrigger& config)
+      : threshold_(config.value()) {}
+
+  bool updateValue(double value) {
+    const bool fired = isFired();
+    value_ = value;
+    return fired != isFired();
+  }
+
+  bool isFired() const { return value_.has_value() && value_ >= threshold_; }
+
+private:
+  const double threshold_;
+  absl::optional<double> value_;
+};
+
+} // namespace
+
+OverloadActionImpl::OverloadActionImpl(
+    const envoy::config::overload::v2alpha::OverloadAction& config) {
+  for (const auto& trigger_config : config.triggers()) {
+    TriggerPtr trigger;
+
+    switch (trigger_config.trigger_oneof_case()) {
+    case envoy::config::overload::v2alpha::Trigger::kThreshold:
+      trigger = std::make_unique<ThresholdTriggerImpl>(trigger_config.threshold());
+      break;
+    default:
+      NOT_REACHED_GCOVR_EXCL_LINE;
+    }
+
+    if (!triggers_.insert(std::make_pair(trigger_config.name(), std::move(trigger))).second) {
+      throw EnvoyException(
+          fmt::format("Duplicate trigger resource for overload action {}", config.name()));
+    }
+  }
+}
+
+bool OverloadActionImpl::updateResourcePressure(const std::string& name, double pressure) {
+  const bool active = isActive();
+
+  auto it = triggers_.find(name);
+  RELEASE_ASSERT(it != triggers_.end(), "");
```

Comment:
These look like program invariants to me, in which case they should probably stay plain ASSERTs. (per https://github.com/envoyproxy/envoy/blob/master/STYLE.md#error-handling).

 I think this is true for most if not all the RELEASE_ASSERTs you have.